### PR TITLE
Nickmale add remove experiment view

### PR DESCRIPTION
**Pull request by NickMale**
Adding view that lists, adds, and removes experiments.

### CODE REVIEW

**Comment by NickMale** [comment]
@alexwaibel you mind handling the conflicts on this one? Some of these I'm not sure which version should be accepted.

**Comment by NickMale** [comment]
@alexwaibel  You mind handling the conflicts on this one? Some of these files I haven't touched and I'm not sure which version should be accepted.

**Comment by alexwaibel** [comment]
> @alexwaibel  You mind handling the conflicts on this one? Some of these files I haven't touched and I'm not sure which version should be accepted.

Yeah I'll have a go at it

File: src/components/experiment/Experiment.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import ExperimentItem from './ExperimentItem';
+
+function generateItems(experimentUrls: Array<string>) {
+  const items: Array<JSX.Element> = [];
+  experimentUrls.forEach(url => {
+    items.push(<ExperimentItem url={url} key={url} />);
+  });
+
+  return items;
+}
```
Comment (alexwaibel) [inline src/components/experiment/Experiment.tsx:18]: May as well use an arrow function like you do for the other methods in this file for consistency.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+function ExperimentItem({ url }: ExperimentItemProps): ReactElement<ExperimentItemProps> {
+  const filenameExpression = new RegExp(`${process.env.REACT_APP_BACKEND_ADDRESS}/static/experiment/(.*)`);
+  const match = filenameExpression.exec(url);
+  const filename = match ? match[1] : 'unknown filename';
+
+  return (
+    <ListGroup.Item>
+      <p>{filename}</p>
+      <Button href={url} className="mr-1">View</Button>
+      <Button onClick={() => deleteExperimentFile(filename)}>Delete</Button>
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:23]: I think it'd look a bit better if we had the buttons right justified on the line with the experiment name.
I'd also prefer the delete button be red.
I think we should have the delete button pop up with a modal asking for confirmation before calling the delete since deleting an experiment is a big deal

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+function ExperimentItem({ url }: ExperimentItemProps): ReactElement<ExperimentItemProps> {
+  const filenameExpression = new RegExp(`${process.env.REACT_APP_BACKEND_ADDRESS}/static/experiment/(.*)`);
+  const match = filenameExpression.exec(url);
+  const filename = match ? match[1] : 'unknown filename';
+
+  return (
+    <ListGroup.Item>
+      <p>{filename}</p>
+      <Button href={url} className="mr-1">View</Button>
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:22]: This view didn't work for me when testing locally.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 4]
```diff
@@ -0,0 +1,28 @@
+import React, { ReactElement } from 'react';
+import ListGroup from 'react-bootstrap/ListGroup';
+import Button from 'react-bootstrap/Button';
+
+interface ExperimentItemProps {
+  url: string
+}
+
+async function deleteExperimentFile(filename: string) {
+  await fetch(`${process.env.REACT_APP_BACKEND_ADDRESS}/experiment/${filename}`, { method: 'delete'});
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:10]: What if the delete call fails?

File: src/components/experiment/__tests__/Experiment.text.tsx [hunk 5]
```diff
@@ -0,0 +1,3 @@
+import React from 'react';
+import { render } from '@testing-library/react';
+import Experiment from '../Experiment';
```
Comment (alexwaibel) [inline src/components/experiment/__tests__/Experiment.text.tsx:3]: You should have some unit tests in here.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 2 above)
Comment (alexwaibel) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:23]: Could we ditch the review button and just make the filename a hyperlink?

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 4 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:10]: It will display the message returned from the delete response

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 2 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:23]: To be honest if we're designing this UI to be as mobile-friendly as we can I think a button is nicer on button than a hyperlink.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 3 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:22]: Any errors in console or anything? Works fine for me on Brave & Chrome.

File: src/components/experiment/__tests__/Experiment.text.tsx (hunk 5 above)
Comment (NickMale) [reply src/components/experiment/__tests__/Experiment.text.tsx]: Added, as well as tests for ExperimentItem

File: src/components/experiment/Experiment.tsx (hunk 1 above)
Comment (NickMale) [reply src/components/experiment/Experiment.tsx]: Refactored in eb6d5da

File: src/components/app/App.tsx [hunk 6]
```diff
@@ -22,6 +23,8 @@ const App: React.FC = () => {
             <Suspense fallback={<div>Loading...</div>}>
               <Switch>
                 <Route exact path="/" component={Home} />
+                <Route exact path="/experiment" component={Experiment} />
+                <Route path="/stimuli" component={Stimuli} />
```
Comment (alexwaibel) [inline src/components/app/App.tsx:27]: Stimuli route shouldn't be duplicated here.

File: src/components/app/App.tsx [hunk 7]
```diff
@@ -22,6 +23,8 @@ const App: React.FC = () => {
             <Suspense fallback={<div>Loading...</div>}>
               <Switch>
                 <Route exact path="/" component={Home} />
+                <Route exact path="/experiment" component={Experiment} />
```
Comment (alexwaibel) [inline src/components/app/App.tsx:26]: exact is not necessary here since we don't have multiple paths starting with `/experiment`

File: src/components/experiment/Experiment.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+  return items;
+};
+
+const Experiment: React.FC = () => {
+  const [hasError, setErrors] = useState(false);
+  const [experimentUrls, setExperimentUrls] = useState([]);
+  const [isUploading, setUploading] = useState(false);
+  const [showToast, setShowToast] = useState(false);
+  const [responseMessage, setResponseMessage] = useState('');
```
Comment (alexwaibel) [inline src/components/experiment/Experiment.tsx:23]: Do we need both showToast and responseMessage? If the response message is null, that's enough info to know not to show the toast. Conversely, if the responseMessage exists we should show a toast. With that in mind, is the showToast state necessary?

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import React, { ReactElement, useState } from 'react';
+import ListGroup from 'react-bootstrap/ListGroup';
+import { Modal, Button } from 'react-bootstrap';
+
+interface ExperimentItemProps {
+  url: string;
+}
+
+async function deleteExperimentFile(filename: string) {
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:9]: This should be an arrow function for consistency.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+  await fetch(
+    `${process.env.REACT_APP_BACKEND_ADDRESS}/experiment/${filename}`,
+    { method: 'delete' }
+  );
+  window.location.reload(false);
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:14]: Why is this necessary?

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+function ExperimentItem({
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:17]: This should also be an arrow function for consistency.

File: src/components/experiment/ExperimentItem/__tests__/ExperimentItem.test.tsx [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    const wrapper = shallow(
+      <ExperimentItem url="http://localhost/static/experiment/some_experiment_url.py" />
+    );
+    await act(async () => {
+      wrapper.find('.delete-button').simulate('click');
+    });
+    expect(wrapper.find(Modal).props()['show']).toBe(true);
+  });
+
+  it('Clicking delete and then delete again issues a fetch', async () => {
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/__tests__/ExperimentItem.test.tsx:31]: Nitpick: this should be phrased more like:
it('issues a deletion call when delete and delete confirmation are clicked")

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+          <Modal.Footer>
+            <Button
+              className="confirm-delete"
+              variant="danger"
+              onClick={() => {
+                deleteExperimentFile(filename);
+                setModalStatus(false);
+              }}
+            >
+              Yes
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:70]: Nit: I'd prefer this said `Delete` rather than `Yes`. I'd also prefer the delete button be on the right.

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+
+  return (
+    <ListGroup.Item className="experiment-item">
+      <span>{filename}</span>
+      <Button
+        href={url}
+        style={{ float: 'right' }}
+        className="mr-1 view-button"
+      >
+        View
```
Comment (alexwaibel) [inline src/components/experiment/ExperimentItem/ExperimentItem.tsx:35]: Nit: I'd prefer the delete button was to the right of the view button.

File: src/components/experiment/Experiment.tsx [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+    setShowToast(true);
+    setSelectedFile({ file: null });
+  };
+
+  return (
+    <Container>
+      {!hasError && (
+        <Row>
+          <Col>
+            <div className="input-group">
```
Comment (alexwaibel) [inline src/components/experiment/Experiment.tsx:71]: We should add a little bit of bottom margin to the upload bar so that there's some whitespace between it and the list of files.

File: src/components/experiment/Experiment.tsx (hunk 8 above)
Comment (NickMale) [reply src/components/experiment/Experiment.tsx:23]: You can't directly assign null to a state, typescript won't allow it. I could likely get around this by passing it { message: null } but that feels less than clean. Up to you though, if that makes more sense I'll change it

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 10 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:14]: The deletion is handled by the ExperimentItem, and the deletion changes only the backend, the Experiment component needs to reload the list of experiments, the simplest solution seemed to be to reload the page. Alternatively, we could do something like [this](https://ourcodeworld.com/articles/read/409/how-to-update-parent-state-from-child-component-in-react) although I'd have to figure out how props work in function components, and it would ultimately have the same effect.

File: src/components/experiment/ExperimentItem/__tests__/ExperimentItem.test.tsx (hunk 12 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/__tests__/ExperimentItem.test.tsx:31]: makes sense

File: src/components/experiment/Experiment.tsx (hunk 15 above)
Comment (NickMale) [reply src/components/experiment/Experiment.tsx:71]: Changed

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 14 above)
Comment (NickMale) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:35]: Changed

File: src/components/experiment/Experiment.tsx (hunk 8 above)
Comment (alexwaibel) [reply src/components/experiment/Experiment.tsx:23]: It's fine as is

File: src/components/experiment/ExperimentItem/ExperimentItem.tsx (hunk 10 above)
Comment (alexwaibel) [reply src/components/experiment/ExperimentItem/ExperimentItem.tsx:14]: That's fine, I was just curious what the purpose was. That makes sense to me as is.
